classes/services/PKPSubmissionFileService.inc.php

```diff
@@ -0,0 +1,837 @@
+									'stageId' => $this->getWorkflowStageId($submissionFile),
+								]
+							),
+						];
+					}
+					$values[$prop] = $revisions;
+					break;
+				case 'size':
+					$values[$prop] = Services::get('file')->fs->getSize($path);
+					break;
+				case 'url':
+					$values[$prop] = $dispatcher->url(
+						$request,
+						ROUTE_COMPONENT,
+						$request->getContext()->getData('urlPath'),
+						'api.file.FileApiHandler',
+						'downloadFile',
+						null,
+						[
+							'submissionFileId' => $submissionFile->getId(),
+							'submissionId' => $submissionFile->getData('submissionId'),
+							'stageId' => $this->getWorkflowStageId($submissionFile),
+						]
+					);
+					break;
+				default:
+					$values[$prop] = $submissionFile->getData($prop);
+					break;
+			}
+		}
+
+		$values = Services::get('schema')->addMissingMultilingualValues(SCHEMA_SUBMISSION_FILE, $values, $request->getContext()->getSupportedFormLocales());
+
+		HookRegistry::call('SubmissionFile::getProperties', [&$values, $submissionFile, $props, $args]);
+
+		ksort($values);
+
+		return $values;
+	}
+
+	/**
+	 * @copydoc \PKP\Services\interfaces\EntityPropertyInterface::getSummaryProperties()
+	 */
+	public function getSummaryProperties($submissionFile, $args = null) {
+		$props = Services::get('schema')->getSummaryProps(SCHEMA_SUBMISSION_FILE);
+
+		return $this->getProperties($submissionFile, $props, $args);
+	}
+
+	/**
+	 * @copydoc \PKP\Services\interfaces\EntityPropertyInterface::getFullProperties()
+	 */
+	public function getFullProperties($submissionFile, $args = null) {
+		$props = Services::get('schema')->getFullProps(SCHEMA_SUBMISSION_FILE);
+
+		return $this->getProperties($submissionFile, $props, $args);
+	}
+
+	/**
+	 * @copydoc \PKP\Services\interfaces\EntityWriteInterface::validate()
+	 */
+	public function validate($action, $props, $allowedLocales, $primaryLocale) {
+		\AppLocale::requireComponents(
+			LOCALE_COMPONENT_PKP_MANAGER,
+			LOCALE_COMPONENT_APP_MANAGER
+		);
+		$schemaService = Services::get('schema');
+
+		import('lib.pkp.classes.validation.ValidatorFactory');
+		$validator = \ValidatorFactory::make(
+			$props,
+			$schemaService->getValidationRules(SCHEMA_SUBMISSION_FILE, $allowedLocales)
+		);
+
+		// Check required fields if we're adding a context
+		\ValidatorFactory::required(
+			$validator,
+			$action,
+			$schemaService->getRequiredProps(SCHEMA_SUBMISSION_FILE),
+			$schemaService->getMultilingualProps(SCHEMA_SUBMISSION_FILE),
+			$allowedLocales,
+			$primaryLocale
+		);
+
+		// Check for input from disallowed locales
+		\ValidatorFactory::allowedLocales($validator, $schemaService->getMultilingualProps(SCHEMA_SUBMISSION_FILE), $allowedLocales);
+
+		// Do not allow the uploaderUserId to be modified
+		if ($action === VALIDATE_ACTION_EDIT) {
+			$validator->after(function($validator) use ($props) {
+				if (!empty($props['uploaderUserId']) && !$validator->errors()->get('uploaderUserId')) {
```

Comment:
There are other invariants, no?

Reply:
Hmm... Properties that really can't be changed, like `id`, are marked as `readOnly` in the schema and will never be changed. 

Something like the `submissionId` can't be changed through the REST API, because it is [set to override any request params](https://github.com/pkp/pkp-lib/pull/6292/files/31b98d3193211ef26c5c4255dea02ec0a8716e6a#diff-f8b67f8fb1d61a986f508962ce9bd322f7211345c396876db656a5841204dfecR268). But I think below that level, it should technically be possible to move a submission file from one submission to another (if someone were to write a CLI tool to merge submissions, for example).

Oh, it looks like `createdAt` should be blocked in the validation method too, when updating. When adding a file, it is stamped automatically. But it does look like it can be edited with an API request so that should probably be prevented in the validation method.

Reply:
I'd have maybe leaned towards making a few more invariant until there's a use case for changing them -- but as long as "dangerous" edits are decently protected, I'm OK with this. (Who is allowed to change a file's submission ID, for example? Only a manager of both the source and target contexts, which are potentially different? An author to both submissions? We're guessing until there's a concrete use case, and any speculative code is likely to get badly maintained.)

Reply:
> We're guessing until there's a concrete use case, and any speculative code is likely to get badly maintained.

I agree, which is why I think it's better to leave such code out of the validate method, which only really checks that the data is valid. Any authorization restrictions need to be applied at the Handler level, which understands where the request came from and what is the action being taken, like we do when editing a file [here](https://github.com/pkp/pkp-lib/pull/6292/files/#diff-f8b67f8fb1d61a986f508962ce9bd322f7211345c396876db656a5841204dfecR355).